### ISSUE TITLE
preprocessing functions for panel data

### ISSUE THREAD

**Issue by geoHeil** [timeseriesAI/tsai#33]
How can I apply the pre-processing functions for panel based time-series data i.e. time series of multiple things (devices)?

Should I set the batch-size to 1 (painfully slow) or is it required to re-code these functions to generate a standardization per panel element or an image for each one?

**Comment by oguiza** [comment]
There are multiple ways to pre-process data using `tsai`: 

1. TSStandardize or TSNormalize based on all train samples: this is the default.
2. TSStandardize or TSNormalize with by_sample=True: this standardizes or normalizes the data by sample (I think this is what you are asking). You can have any batch size.
3. TSStandardize or TSNormalize with by_channel=True: this standardizes or normalizes the data by channel based on the train set.
4. TSStandardize or TSNormalize with by_sample=True, by_channel=True: this standardizes or normalizes each channel of each sample independently of the rest. 

I've tested the different approaches with different datasets, and you'll get different results. It's good to try different approaches to see as performance seems to be data-dependent.

**Comment by TKassis** [comment]
Just to clarify, TSStandarize and TSNormalize work only on a batch level, correct? Is it possible to use them for the entire dataset? I'm particularly interested in normalizing my data but worried that a single batch might not capture the extremes of the values for the variables.

**Comment by oguiza** [comment]
> Just to clarify, TSStandarize and TSNormalize work only on a batch level, correct? Is it possible to use them for the entire dataset? I'm particularly interested in normalizing my data but worried that a single batch might not capture the extremes of the values for the variables.

Hi @TKassis, 
In that case, you may want to calculate the mean and std (or min and max for TSNormalize) using all the train samples, and pass them as kwargs. If you pass mean and std or mina and max, the rest of the arguments are not used, and the batch_tfm won't be set up using a single batch.

**Comment by oguiza** [comment]
> Just to clarify, TSStandarize and TSNormalize work only on a batch level, correct? Is it possible to use them for the entire dataset? I'm particularly interested in normalizing my data but worried that a single batch might not capture the extremes of the values for the variables.

Hi @TKassis,
I’ve been thinking about your comment. I’ve faced the same issue sometimes. So I have added a new argument in both TSStandardize and TSNormalize called use_single_batch. By default it’s set to True. But if you set it to False, the scaler will use the entire training set to calculate the necessary stats. I hope this will meet your needs.

**Comment by TKassis** [comment]
This is perfect! I'll try it out. I ended up calculating the stats before the training loop and normalized accordingly.
